lib/api/apiUtils/object/locationStorageCheck.js

```diff
@@ -0,0 +1,54 @@
+const { errors } = require('arsenal');
+
+const { config } = require('../../../Config');
+const { listLocationMetric, pushLocationMetric } =
+    require('../../../utapi/utilities');
+
+function _gbToBytes(gb) {
+    return gb * 1024 * 1024 * 1024;
+}
+
+/**
+ *
+ * @param {number} newObjSize - new size to check against quota in bytes
+ * @param {number} prevObjSize - old size of obj if existed, else null
+ * @param {boolean} isDeleteMarker - true if delete marker
+ * @param {string} location - name of location to check quota
+ * @param {object} log - werelogs logger
+ * @param {function} cb - callback function
+ * @return {undefined}
+ */
+function locationStorageCheck(newObjSize, prevObjSize, isDeleteMarker, location,
+log, cb) {
+    const sizeLimitGB = config.locationConstraints[location].
+        details.sizeLimitGB;
+    const sizeDiff = newObjSize - prevObjSize;
+
+    if (isDeleteMarker || sizeDiff === 0 || sizeLimitGB === undefined) {
+        return cb();
+    }
+    // no need to list location metric, since it should be decreased
+    if (sizeDiff < 0) {
+        return pushLocationMetric(location, sizeDiff, log, cb);
+    }
+    return listLocationMetric(location, log, (err, bytesStored) => {
+        if (err) {
+            log.error(`Error listing metrics from Utapi: ${err.message}`);
+            return cb(err);
+        }
+        const newStorageSize = bytesStored + sizeDiff;
+        const sizeLimitBytes = _gbToBytes(sizeLimitGB);
+        if (sizeLimitBytes < newStorageSize) {
+            return cb(errors.AccessDenied.customizeDescription(
+                `The assigned storage space limit for location ${location} ` +
+                'will be exceeded'));
+        }
+        if ((newStorageSize / sizeLimitBytes) > 0.8) {
+            log.warn(`${location} location storage space is above 80% ` +
+                'capacity');
+        }
+        return pushLocationMetric(location, sizeDiff, log, cb);
```

Comment:
I'm wondering if the fact that we have the `pushLocationMetric` inside the data backend (`_retryDelete`) will not have the effect of carrying the decrement twice (here and there)?

Overall, I haven't dug how or if this is the right thing to do, but it seems to me the logical answer is to increment the location metric in the data backend itself (`data.put`), so that it's in par with the decrement in `data.delete`, it's much easier to reason about and much less likely to forget to update some code paths (now or in the future). Again if there's a valid reason to do the increments outside the data wrapper layer, please disregard, just thinking out loud.

Reply:
Yeah, I originally put it in here since there is a path in the objectDelete API that goes through the `createAndStoreObject` file, but since that case is only for delete markers (which don't affect size count), it makes more sense to put the metric check in the wrapper